redis/tests/test_cache.rs

```diff
@@ -0,0 +1,307 @@
+        // It should get value once and miss then it should use cache.
+        assert_eq!(s.hit, 14);
+        assert_eq!(s.miss, 1);
+        // 1 GET + 1 PTTL
+        assert_eq!(s.sent_command_count, 2);
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_guards_and_nil_values() {
+    // Testing guard will work even with Nil Value is returned from server
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+        let mut v = vec![];
+        for _ in 0..5 {
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<u32> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<String> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+            let mut c = con.clone();
+            v.push(tokio::spawn(async move {
+                let x: Option<u8> = c.get("KEY").await.unwrap();
+                assert_eq!(x, None);
+            }));
+        }
+        futures::future::join_all(v).await;
+        let s = con.get_cache_statistics();
+        // It should get value once and miss then it should use cache.
+        assert_eq!(s.hit, 14);
+        assert_eq!(s.miss, 1);
+        // 1 GET + 1 PTTL
+        assert_eq!(s.sent_command_count, 2);
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_verify() {
+    // Verifies task cancellation of request with same key shouldn't affect other requests to same key.
+    use redis::ProtocolVersion;
+
+    let ctx = TestContext::new();
+    if ctx.protocol == ProtocolVersion::RESP2 {
+        return;
+    }
+
+    block_on_all(async move {
+        let mut con = ctx
+            .multiplexed_async_connection_tokio_with_cache_config(CacheConfig::default())
+            .await?;
+        let c2 = con.clone();
+        let join_handle = tokio::spawn(async move {
+            let con = c2.clone();
+            let mut v = vec![];
+            for i in 0..100000 {
+                let mut c = con.clone();
+                v.push(async move {
+                    let x: RedisFuture<Option<u32>> = c.get("KEY");
+                    let _ = tokio::time::timeout(Duration::from_micros(i * 10), x).await;
+                })
+            }
+            futures::future::join_all(v).await;
+        });
+        for _ in 0..10000 {
+            let _: Option<u32> = con.get("KEY").await.unwrap();
+        }
+        join_handle.await.unwrap();
+        Ok::<_, RedisError>(())
+    })
+    .unwrap();
+}
+
+#[test]
+fn test_cache_mget_verify() {
```

Comment:
I'm having a hard time understanding the purpose of all of the tests. The names are very terse, and require knowledge of the internal implementation in order to understand what they're actually testing.

Reply:
I will add more documentation and make them clearer.